src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,244 @@
+     * @param httpStack HTTP stack to be used
+     * @param pool a buffer pool that improves GC performance in copy operations
+     */
+    private BasicAsyncNetwork(AsyncHttpStack httpStack, ByteArrayPool pool) {
+        super(httpStack);
+        mPool = pool;
+    }
+
+    /* Method to be called after a successful network request */
+    private void onRequestSucceeded(
+            final Request<?> request,
+            final long requestStartMs,
+            final HttpResponse httpResponse,
+            final OnRequestComplete callback) {
+        final int statusCode = httpResponse.getStatusCode();
+        final List<Header> responseHeaders = httpResponse.getHeaders();
+        // Handle cache validation.
+        if (statusCode == HttpURLConnection.HTTP_NOT_MODIFIED) {
+            long requestDuration = SystemClock.elapsedRealtime() - requestStartMs;
+            callback.onSuccess(
+                    NetworkUtility.getNotModifiedNetworkResponse(
+                            request, requestDuration, responseHeaders));
+            return;
+        }
+
+        byte[] responseContents;
+        if (httpResponse.getContentLength() == -1) {
+            // Add 0 byte response as a way of honestly representing a
+            // no-content request.
+            responseContents = new byte[0];
+        } else {
+            responseContents = httpResponse.getContentBytes();
+        }
+
+        if (responseContents != null) {
+            onResponseRead(
+                    requestStartMs,
+                    statusCode,
+                    httpResponse,
+                    request,
+                    callback,
+                    responseHeaders,
+                    responseContents);
+            return;
+        }
+
+        // The underlying AsyncHttpStack does not support asynchronous reading of the response into
+        // a byte array, so we need to submit a blocking task to copy the response from the
+        // InputStream instead.
+        final InputStream inputStream = httpResponse.getContent();
+        Runnable run =
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        byte[] finalResponseContents;
+                        try {
+                            finalResponseContents =
+                                    NetworkUtility.inputStreamToBytes(
+                                            inputStream, httpResponse.getContentLength(), mPool);
+                        } catch (IOException e) {
+                            onRequestFailed(
+                                    request, callback, e, requestStartMs, httpResponse, null);
+                            return;
+                        } catch (ServerError serverError) {
+                            // This should never happen since we already check if inputStream is
+                            // null
+                            throw new RuntimeException(serverError);
```

Comment:
perhaps IllegalStateException - or consider changing inputStreamToBytes not to throw ServerError and moving that logic into BasicNetwork if it's unused in BasicAsyncNetwork's flow.

Reply:
I think handling it in BasicNetwork is better, and then not dealing with the ServerError in BasicAsyncNetwork.